### ISSUE TITLE
add support for hardening user_project_wrapper

### ISSUE THREAD

**Issue by proppy** [The-OpenROAD-Project/OpenLane-MPW-CI#108]
Currently we CI seems to only harden the user project macro, it would be nice to also harden the `user_project_wrapper` so that we can also catch regression over the integration with caravel.

**Comment by maliberty** [comment]
Also the effective area is smaller.  Anton had to remove the FPU from microwatt to make it fit.

**Comment by tspyrou** [comment]
@vijayank88 and @vvbandeira would it be hard to enhance this to also do what @proppy asks?

**Comment by vijayank88** [comment]
30% designs already added and running for `user_project_wrapper`.
Most of designs fails with DRC or LVS for wrapper designs. https://jenkins.openroad.tools/blue/organizations/jenkins/OpenLane-MPW-CI-Public%2Funstable-small-1/detail/PR-109/1/pipeline

Those who forked from `efabless/caravel_user_project ` we're not able file github issue about design failures.

@dralabeing  and @vvbandeira  working on collecting user email ID to send automatic notification for design failures, so user can fix and update the repo.

**Comment by vijayank88** [comment]
@tspyrou @proppy 
With existing `user_project_wrapper` itself we can able to figure out OpenLane flow/scripts/PDK issues we're able to find and reporting back for the fixes. Do you want still to add all remaining `user_project_wrapper` shuttle designs in CI?

CI failures and respective issue reported with label `MPW CI` https://github.com/The-OpenROAD-Project/OpenLane/issues?q=label%3A%22MPW+CI%22+is%3Aclosed

**Comment by proppy** [comment]
> 30% designs already added and running for user_project_wrapper.

can you give an example of such a design?

If I'm looking at a project like:
https://github.com/The-OpenROAD-Project/OpenLane-MPW-CI/blob/ae47acfbe9ebd14959cf20a55875878f299e25a4/scripts/spectrometer_hyperspace_mpw7/config.tcl#L7
It seems that it *only* harden the wrapper and so will most likely reuse the pre-built macro check-in in the repo (which would be tied to sky130).

Do we have a mechanism to run the flow 2x per design?
- 1x for the user macro
- 1x for the user_project_wrapper
-

**Comment by proppy** [comment]
> we can able to figure out OpenLane flow/scripts/PDK issues, able to find and reporting back for the fixes

For now, I'm mostly after getting some coverage on the number of design that can be ported with little to no change to `PDK=gf180mcuC`, but after we got "some designs" to pass the GF flow I do agree that it will be nice to report those as regression if we find out a given tool commit make them fail.

**Comment by proppy** [comment]
> Here's list of user_project_wrapper design for sky130 pdk

Yes, but it seems to me (if I'm reading the `run_design.sh` script correctly, that it would *only* harden the wrapper for those project (and not the actual project macro)

**Comment by vijayank88** [comment]
Yes, you're right. Its look for `openlane/design_to_harden/config.tcl` and run OpenLane flow for the same like `caravel_user_project`.
If design fails at any stage, respective changes applied through `scripts/design_name/setup_flow.sh` to complete the flow.

**Comment by proppy** [comment]
> run OpenLane flow for the same like caravel_user_project.

yep, but as point in https://caravel-user-project.readthedocs.io/en/latest/#quickstart, caravel_user_project requires to run ```
make user_proj_example
make user_project_wrapper
```
to go thru the whole flow.

So I believe we miss an extra `make` call for those project here:
https://github.com/The-OpenROAD-Project/OpenLane-MPW-CI/blob/main/scripts/run-design.sh#L55

**Comment by vijayank88** [comment]
@proppy 
This CI setup made to pass some additional configuration through `run_design.sh` as not all designs works 100% at our end.
There are many configuration update from MPW-1 to MPW-7. So we made this way. 
@vvbandeira you want to any additional information on this?

**Comment by proppy** [comment]
> This CI setup made to pass some additional configuration through run_design.sh as not all designs works 100% at our end.

I'm not really worried about `setup_flow.sh `, I think maintaining a layer of patches is a fine thing to do from the point of view of the CI (as long as you also try to submit them back to project for integration, or use it as a hint to document or fix backward compatibility issues in the tooling).

To be more precise: what I'm worried here is that, apparently for those projects:
- `run_design.sh` only runs `flow.tcl` on `user_project_wrapper`
- `run_design.sh` doesn't re-runs `flow.tcl` on the main user-provided project configuration to create the main user-provided project macro, and just end up reusing whatever is in the `gds/` directory of the project submodules.

Maybe @QuantamHD or @mithro could jump in to try to clarify what I'm trying to express here.

**Comment by dralabeing** [comment]
@maliberty Can you elaborate on the issue faced by Anton? Was he unable to fit FPU on  microwatt on a later shuttle that had updates to Caravel (user area)? Why is this a CI issue?

**Comment by maliberty** [comment]
It isn't a CI issue.  It is case where you can't just use a skywater design unmodified on gf180.  It is an effective area limitation.

**Comment by dralabeing** [comment]
@maliberty Thanks... got it.
